src/testutil/mod.rs

```diff
@@ -0,0 +1,276 @@
+        ) -> Self::Future {
+             unreachable!("no async builtins should be called in this test")
+        }
+    }
+
+    let tokd = tokenize(contents, true).unwrap();
+    let mut forth = AsyncLBForth::from_params(tokd.settings, (), Forth::FULL_BUILTINS, TestAsyncDispatcher);
+    async_blockon_runtest_with(contents, &mut forth.forth);
+}
+
+/// Like `async_blockon_runtest`, but with provided context + dispatcher
+#[cfg(feature = "async")]
+pub fn async_blockon_runtest_with_dispatcher<T, D>(contents: &str, context: T, dispatcher: D)
+where
+    T: 'static,
+    D: for<'forth> crate::dictionary::AsyncBuiltins<'forth, T>,
+{
+    use crate::leakbox::AsyncLBForth;
+
+    let tokd = tokenize(contents, true).unwrap();
+    let mut forth = AsyncLBForth::from_params(tokd.settings, context, Forth::FULL_BUILTINS, dispatcher);
+    async_blockon_runtest_with(contents, &mut forth.forth);
+}
+
+/// Like `async_blockon_runtest`, but with provided async vm
+#[cfg(feature = "async")]
+pub fn async_blockon_runtest_with<T, D>(contents: &str, forth: &mut crate::AsyncForth<T, D>)
+where
+    T: 'static,
+    D: for<'forth> crate::dictionary::AsyncBuiltins<'forth, T>,
+{
+    let tokd = tokenize(contents, false).unwrap();
+    for Step { input, output: outcome } in tokd.steps.iter() {
+        forth.input_mut().fill(&input).unwrap();
+        let res = futures::executor::block_on(forth.process_line());
+        check_output(res, outcome, forth.output().as_str());
+        forth.output_mut().clear();
+    }
+}
+
+fn check_output(res: Result<(), Error>, outcome: &Outcome, output: &str) {
+    match (res, outcome) {
+        (Ok(()), Outcome::OkAnyOutput) => {}
+        (Ok(()), Outcome::OkWithOutput(exp)) => {
+            let act_lines = output.lines().collect::<Vec<&str>>();
+            assert_eq!(act_lines.len(), exp.len());
+            act_lines.iter().zip(exp.iter()).for_each(|(a, e)| {
+                assert_eq!(a.trim_end(), e.trim_end());
+            })
+        }
+        (Err(_e), Outcome::FatalError) => {}
+        (res, exp) => {
+            eprintln!("Error!");
+            eprintln!("Expected: {exp:?}");
+            eprintln!("Got: {res:?}");
+            if res.is_ok() {
+                eprintln!("Output:\n{}", output);
+            }
+            panic!();
+        }
+    }
+}
+
+// Runs the given steps against the given forth VM.
+//
+// Panics on any mismatch
+fn blocking_steps_with<T>(steps: &[Step], forth: &mut Forth<T>) {
+    for Step { input, output: outcome } in steps.into_iter() {
+        forth.input.fill(&input).unwrap();
+        let res = forth.process_line();
+        check_output(res, outcome, forth.output.as_str());
+        forth.output.clear();
+    }
+}
+
+#[derive(Debug)]
+pub enum Outcome {
+    OkAnyOutput,
+    OkWithOutput(Vec<String>),
+    FatalError,
+}
+
+#[derive(Debug)]
+pub struct Step {
+    pub input: String,
+    pub output: Outcome,
+}
+
+#[derive(Default, Debug)]
+pub struct Tokenized {
```

Comment:
nit: do any of these need to be `pub`? AFAICT, they aren't exposed in any of the public API functions' signatures?
```suggestion
enum Outcome {
    OkAnyOutput,
    OkWithOutput(Vec<String>),
    FatalError,
}

#[derive(Debug)]
struct Step {
    pub input: String,
    pub output: Outcome,
}

#[derive(Default, Debug)]
struct Tokenized {
```